File: src/ir/serialized/serialized-expression-statement-head.ts
import {
  boolean,
  lazy,
  object,
  pipe,
  readonly,
  string,
  variant,
  type GenericSchema,
  type InferOutput,
} from "valibot";

import { filledString$, type FilledString } from "../../util/filled-string.js";
import { span$, type Span } from "../primitive/span.js";
import {
  identifier$,
  member$,
  call$,
  new$,
  await$,
  assign$,
  update$,
  elided$,
  raw$,
} from "../reference/expression-statement-head-kind.js";

// Serialized counterpart of `src/ir/reference/expression-statement-head.ts`.
// Identical shape except the `raw` leaf carries `Span` (line/column/offset)
// instead of bare offsets, matching the convention used elsewhere in the
// serialized IR.
//
// The schema is recursive (member.object, call.callee, new.callee,
// await.argument all reference SerializedHeadExpression). valibot's lazy()
// breaks the cycle so InferOutput can resolve the type.
// Serialized counterpart of HeadOperand: the per-side span info that
// keeps source-position fidelity for assign / update operands (matches
// the convention used by raw heads: Span carries line+column+offset).
const serializedHeadOperand$: GenericSchema<unknown, SerializedHeadOperand> =
  lazy(() =>
    pipe(
      object({
        head: serializedHeadExpression$,
        startSpan: span$,
        endSpan: span$,
      }),
      readonly(),
    ),
  );

export const serializedHeadExpression$: GenericSchema<
  unknown,
  SerializedHeadExpression
> = lazy(() =>
  variant("kind", [
    pipe(
      object({
        kind: identifier$,
        name: filledString$,
      }),
      readonly(),
    ),
    pipe(
      object({
        kind: member$,
        object: serializedHeadExpression$,
        property: filledString$,
      }),
      readonly(),
    ),
    pipe(
      object({
        kind: call$,
        callee: serializedHeadExpression$,
      }),
      readonly(),
    ),
    pipe(
      object({
        kind: new$,
        callee: serializedHeadExpression$,
      }),
      readonly(),
    ),
    pipe(
      object({
        kind: await$,
        argument: serializedHeadExpression$,
      }),
      readonly(),
    ),
    pipe(
      object({
        kind: assign$,
        operator: string(),
        left: serializedHeadOperand$,
        right: serializedHeadOperand$,
      }),
      readonly(),
    ),
    pipe(
      object({
        kind: update$,
        operator: string(),
        prefix: boolean(),
        argument: serializedHeadOperand$,
      }),
      readonly(),
    ),
    pipe(object({ kind: elided$ }), readonly()),
    pipe(
      object({
        kind: raw$,
        startSpan: span$,
        endSpan: span$,
      }),
      readonly(),
    ),
  ]),
);

export type SerializedHeadOperand = Readonly<{
  head: SerializedHeadExpression;
  startSpan: Span;
  endSpan: Span;
}>;

export type SerializedHeadExpression =
  | Readonly<{
      kind: InferOutput<typeof identifier$>;
      name: FilledString;
    }>
  | Readonly<{
      kind: InferOutput<typeof member$>;
      object: SerializedHeadExpression;
      property: FilledString;
    }>
  | Readonly<{
      kind: InferOutput<typeof call$>;
      callee: SerializedHeadExpression;
    }>
  | Readonly<{
      kind: InferOutput<typeof new$>;
      callee: SerializedHeadExpression;
    }>
  | Readonly<{
      kind: InferOutput<typeof await$>;
      argument: SerializedHeadExpression;
    }>
  | Readonly<{
      kind: InferOutput<typeof assign$>;
      operator: string;
      left: SerializedHeadOperand;
      right: SerializedHeadOperand;
    }>
  | Readonly<{
      kind: InferOutput<typeof update$>;
      operator: string;
      prefix: boolean;
      argument: SerializedHeadOperand;
    }>
  | Readonly<{ kind: InferOutput<typeof elided$> }>
  | Readonly<{
      kind: InferOutput<typeof raw$>;
      startSpan: Span;
      endSpan: Span;
    }>;

File: crates/unsnarl-ir/src/serialized/serialized_expression_statement_head.rs
//! Serialized counterpart of `HeadExpression`.
//!
//! Identical shape to the internal head expression except the `Raw`
//! leaf carries `Span` (line/column/offset) instead of bare offsets,
//! and `Assign` / `Update` operands likewise carry spans rather than
//! raw offsets.

use serde::Serialize;
use unsnarl_oxc_parity::{AssignOperator, UpdateOperator};

use crate::non_empty::assert_non_empty;
use crate::primitive::Span;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SerializedHeadOperand {
    pub head: SerializedHeadExpression,
    pub start_span: Span,
    pub end_span: Span,
}

#[derive(Serialize)]
#[serde(tag = "kind", rename_all = "lowercase")]
pub enum SerializedHeadExpression {
    Identifier {
        name: String,
    },
    Member {
        object: Box<SerializedHeadExpression>,
        property: String,
    },
    /// `start_span` / `end_span` carry the **enclosing
    /// CallExpression**'s span. They are emitted so consumers
    /// (notably the visual-graph builder's callback-arg labeller)
    /// can match a call node inside a chained expression
    /// (`a.b().c(cb)`) where every nested `CallExpression` shares
    /// the same `start.offset` with the chain root.
    #[serde(rename_all = "camelCase")]
    Call {
        callee: Box<SerializedHeadExpression>,
        start_span: Span,
        end_span: Span,
    },
    /// `start_span` / `end_span` carry the **enclosing
    /// NewExpression**'s span. Same identification rationale as
    /// [`Self::Call`].
    #[serde(rename_all = "camelCase")]
    New {
        callee: Box<SerializedHeadExpression>,
        start_span: Span,
        end_span: Span,
    },
    Await {
        argument: Box<SerializedHeadExpression>,
    },
    Assign {
        operator: AssignOperator,
        left: Box<SerializedHeadOperand>,
        right: Box<SerializedHeadOperand>,
    },
    Update {
        operator: UpdateOperator,
        prefix: bool,
        argument: Box<SerializedHeadOperand>,
    },
    Elided,
    #[serde(rename_all = "camelCase")]
    Raw {
        start_span: Span,
        end_span: Span,
    },
}

impl SerializedHeadExpression {
    pub fn identifier(name: String) -> Self {
        assert_non_empty(&name, "SerializedHeadExpression::Identifier.name");
        Self::Identifier { name }
    }

    pub fn member(object: SerializedHeadExpression, property: String) -> Self {
        assert_non_empty(&property, "SerializedHeadExpression::Member.property");
        Self::Member {
            object: Box::new(object),
            property,
        }
    }
}
